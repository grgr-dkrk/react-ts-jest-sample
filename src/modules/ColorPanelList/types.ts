import * as Actions from './actions';
import { ActionType } from 'redux-actions-type';

/**
 * ColorPanel
 */
export type ColorPanelType = {
  id: number;
  label?: string;
  colorCode: string;
  published?: string;
};

export type DataType = Pick<ColorPanelType, 'label' | 'colorCode'>;

/**
 * State
 */
export type StateType = {
  items: ColorPanelType[];
};

/**
 * ActionTypes
 */
export type ActionTypes = ActionType<typeof Actions>;
