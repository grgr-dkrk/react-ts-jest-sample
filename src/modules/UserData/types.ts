import * as Actions from './actions';
import { ActionType } from 'redux-actions-type';

export type UserStateType = {
  name: string;
  token: string;
};

export type StateType = {
  data: UserStateType | null;
};

export type ActionTypes = ActionType<typeof Actions>;
