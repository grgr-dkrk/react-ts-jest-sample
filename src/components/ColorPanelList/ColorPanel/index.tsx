import * as React from 'react';
import { ColorPanelType } from '@/modules/ColorPanelList/types';

type ColorPanelProps = ColorPanelType;

const ColorPanel = (props: ColorPanelProps) => {
  return (
    <li>
      {props.label || props.colorCode}
      <span
        role="presentation"
        style={{
          display: 'block',
          backgroundColor: props.colorCode,
          height: 60,
        }}
      ></span>
    </li>
  );
};

export default ColorPanel;
