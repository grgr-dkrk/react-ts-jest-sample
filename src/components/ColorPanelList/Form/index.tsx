import * as React from 'react';
import Input from './Input';
import SubmitButton from './Button';
import { ColorPanelType } from '@/modules/ColorPanelList/types';

type FormProps = {
  handleSubmit: (data: ColorPanelType) => void;
  handleFetchData: () => void;
  itemCount: number;
};

const Form: React.FC<FormProps> = props => {
  const [label, setLabel] = React.useState('');
  const [colorCode, setColorCode] = React.useState('#000');

  const changeLabel = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setLabel(e.target.value);
    },
    []
  );
  const changeColorCode = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setColorCode(e.target.value);
    },
    []
  );
  const submit = React.useCallback(() => {
    props.handleSubmit({ id: props.itemCount, label, colorCode });
  }, [props.itemCount, label, colorCode]);

  const fetchData = React.useCallback(() => {
    props.handleFetchData();
  }, []);

  return (
    <form>
      <Input inputId="title" label="名前" handleChange={changeLabel} />
      <Input
        inputId="colorCode"
        label="色"
        type="color"
        handleChange={changeColorCode}
      />
      <SubmitButton handleClick={submit}>追加する</SubmitButton>
      <SubmitButton handleClick={fetchData}>
        本日のラッキーカラーを追加
      </SubmitButton>
    </form>
  );
};

export default Form;
