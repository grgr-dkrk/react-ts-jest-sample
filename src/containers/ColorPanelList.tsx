import ColorPanelList from '@/components/ColorPanelList/';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { AppState } from '@/store';
import * as Actions from '@/modules/ColorPanelList/actions';
import {
  ColorPanelType,
  ActionTypes,
  DataType,
} from '@/modules/ColorPanelList/types';

const mapStateToProps = (state: AppState) => {
  return {
    items: state.ColorPanelList.items,
    user: state.User.data,
  };
};

export type MapDispatchTypes = {
  add: (data: ColorPanelType) => void;
  remove: (id: number) => void;
  // fetch: () => void;
};

/**
 * Dispatch
 */
const mapDispatchToProps = (
  dispatch: Dispatch<ActionTypes>
): MapDispatchTypes => {
  return {
    add: data => dispatch(Actions.addColorPanel(data)),
    remove: id => dispatch(Actions.removeColorPanel(id)),
    // @ts-ignore
    // fetch: () => dispatch(Actions.fetchLuckyColorDataWithPromise()),
  };
};

/**
 * Connect
 */
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ColorPanelList);
