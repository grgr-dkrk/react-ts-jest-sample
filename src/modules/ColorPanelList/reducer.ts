import { StateType, ActionTypes } from '@/modules/ColorPanelList/types';

/**
 * State
 */
export const initialState: StateType = {
  items: [],
};

/**
 * Reducer
 */
export default function reducer(
  state: StateType = initialState,
  action: ActionTypes
): StateType {
  switch (action.type) {
    case 'addColorPanel':
      return {
        ...state,
        items: [...state.items, action.payload],
      };
    case 'removeColorPanel':
      return {
        ...state,
        items: state.items.filter(item => {
          return item.id !== action.payload;
        }),
      };
    default:
      return state;
  }
}
