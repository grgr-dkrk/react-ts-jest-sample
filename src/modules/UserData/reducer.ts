import { StateType, ActionTypes } from './types';

/**
 * State
 */
export const initialState: StateType = {
  data: {
    name: 'dkrk',
    token: '',
  },
};

/**
 * Reducer
 */
export default function reducer(
  state: StateType = initialState,
  action: ActionTypes
) {
  switch (action.type) {
    case 'setUserData':
      return {
        ...state,
      };
    default:
      return state;
  }
}
