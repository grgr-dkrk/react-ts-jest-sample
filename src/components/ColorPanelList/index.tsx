import * as React from 'react';
import { ColorPanelType } from '@/modules/ColorPanelList/types';
import { MapDispatchTypes } from '@/containers/ColorPanelList';
import ColorPanel from './ColorPanel';
import Form from './Form';
type ColorPanelListProps = {
  items: ColorPanelType[];
} & MapDispatchTypes;

const ColorPanelList: React.FC<ColorPanelListProps> = props => {
  return (
    <div>
      <Form
        handleSubmit={props.add}
        handleFetchData={() => {}}
        itemCount={props.items.length}
      />
      <ul>
        {props.items.map(item => (
          <ColorPanel
            id={item.id}
            label={item.label || ''}
            colorCode={item.colorCode}
            key={item.id}
          />
        ))}
      </ul>
    </div>
  );
};

export default ColorPanelList;
