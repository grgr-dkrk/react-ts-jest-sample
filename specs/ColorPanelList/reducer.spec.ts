import reducer, { initialState } from '@/modules/ColorPanelList/reducer';
import {
  addColorPanel,
  removeColorPanel,
} from '@/modules/ColorPanelList/actions';
import { StateType } from '@/modules/ColorPanelList/types';

describe('ColorPanelListのReducer', () => {
  // add
  describe('add', () => {
    let state: StateType;
    state = reducer(
      initialState,
      addColorPanel({ id: 0, label: '漆黒の黒', colorCode: '#000' })
    );
    it('id', () => {
      expect(state.items[0].id).toBe(0);
    });
    it('label', () => {
      expect(state.items[0].label).toBe('漆黒の黒');
    });
    it('colorCode', () => {
      expect(state.items[0].colorCode).toBe('#000');
    });
  });

  // remove
  describe('remove', () => {
    let state: StateType;
    state = reducer(initialState, addColorPanel({ id: 10, colorCode: '#000' }));
    it('何も残らない', () => {
      let newState = reducer(state, removeColorPanel(10));
      expect(newState.items).toHaveLength(0);
    });
  });

  // default
  describe('対象外のAction', () => {
    let state: StateType;
    // @ts-ignore (型チェックを回避するため)
    state = reducer(undefined, {});
    it('initialStateが返る', () => {
      expect(state).toEqual(initialState);
    });
  });
});
