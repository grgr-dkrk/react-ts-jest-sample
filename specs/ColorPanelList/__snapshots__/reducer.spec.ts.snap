// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPanelListのReducer addAction 1`] = `
Object {
  "items": Array [
    Object {
      "colorCode": "#000",
      "id": 10,
      "label": "漆黒の黒",
    },
  ],
}
`;

exports[`ColorPanelListのReducer removeAction 1`] = `
Object {
  "items": Array [],
}
`;
